### ISSUE TITLE
How i can add a Validator in runtime?

### ISSUE THREAD

**Issue by astrO1** [joanpablo/reactive_forms#129]
How can i change a FormControl validation if i already set it?

**Comment by joanpablo** [comment]
Hi @astrO1,

Could you please give me an example of what you are trying to achieve?

**Comment by fchiumeo** [comment]
I think it refers to this functionality #109

**Comment by joanpablo** [comment]
Hi @astrO1, @fchiumeo,

I would like to know about an example (use case) and the reasons behind this requirements. Some other times in the past other has ask me to include this functionality, but when I ask why is necessary this feature, and they show me, results that the solution was never add a validator dynamically, and the solution was just has the right declaration of Forms and a better understanding of Reactive Forms.

Don't make me wrong, maybe you will need this requirement but I need to be sure that is the only solution you need before implemented it. That's why I would like that you argue your needs and help me understand what you are trying to achieve so I can help you better.

Thanks, best regards.

**Comment by fchiumeo** [comment]
Hi @joanpablo 

I have an item search form that brings me the data from an API service, i have 2 discount fields that have a minimum and maximum based on the quantity
example: quantity: 5, minimum: 0%, maximum: 15%
quantity: 20, minimum: 30%, maximum: 75%
and these values are dynamic according to item and quantity and the best way I found is to set them when the quantity is changed so I need to validate them at runtime

**Comment by piuskamil** [comment]
I think u can't add validator at runtime. Workaround for that is addError I guess. 
In my case I have created simple validation function and manually set error

**Comment by joanpablo** [comment]
Hi @fchiumeo,

could you please show me your FormGroup definition? And give me an example of the data you receive from the server and the validation you need to perform? Sorry, a still don't understand very well why you need to add a validator dynamically.

Don't make me wrong I want to help you to complete your feature. In the past, I also have a similar response with a minimum a maximum, and I needed to validate the field, but I didn't require to add dynamically validators. Maybe this is not the same case that you are having. That's why I need to fully understand your needs. I want to help you. Maybe at the end, you will need a dynamical validator but I need to fully understand why before including this complexity in the package code.

I hope you understand.

Thanks.

**Comment by joanpablo** [comment]
The other reason I need to fully understand the needs is that I need to know how to implement the solution. For example, maybe the solution is not to add/remove dynamically a validator, but instead, define a validator that will only be triggered under certain conditions (you can define a condition that evaluates each time something in the form changes or simply a static condition).

**Comment by joanpablo** [comment]
Or maybe you just need to create a simple **FormGroup** validator that evaluates the value of a control and **addError** or remove an error from another control in the same form depending on your validation logic. That will definitely solve your issue.

**Comment by piuskamil** [comment]
@joanpablo would be nice to have here addValidator functionality like in angular implementation.
For example I want to add required validator to field when I check a checkbox.
I know I can achieve it by addError but it's sounds like workaround.

**Comment by joanpablo** [comment]
> @joanpablo would be nice to have here addValidator functionality like in angular implementation.
> For example I want to add required validator to field when I check a checkbox.
> I know I can achieve it by addError but it's sounds like workaround.

Hi @piuskamil,

I know that adding dynamically a validator will solve the issue, but will include a great complexity, and maybe the solution can be achieve by other way for example:

Angular is more imperative, and Flutter is more declarative.
To add a validator you need anyway to add a listener to the **valueChanges** of the control you want to evaluate, and inside of that listener you must add or *remove* (we will need this one too) a validator. Actually, instead of that, you can create a simple validator at the Group level (that already is listening to all changes in the FormGroup so you don't need to add listeners to *valueChanges* of a particular control) and actually instead of adding/removing a validator you can simple add/remove an error (is almost the same, in principle), and is more natural and less complex.

Maybe you need to validate using values from several controls, so you will need to subscribe to **valueChanges** of several controls, instead of that, just add a validator to a FormGroup (in the declaration) that already receives the data of all controls, and react to changes in all controls.

**Comment by joanpablo** [comment]
Don't make me wrong, I'm not denying the functionality of adding a validator. Maybe in the future, we will need it. But I need to fully understand why and when it is needed. I want to focus on the problem to think about all the solutions we can create (or the ones that we already have) to bring high-quality features. With the help of all of you :)

**Comment by joanpablo** [comment]
In my last project I had a similar problem. I received a collection a data like this:

```json
{
   "USD": {"min": 100, "max": 1000},
   "EUR": {"min": 50, "max": 500}
}
```

And depending on the **Currency** the user had selected (in a ReactiveDropDownField) I needed to validates the **amount** the user typed in a ReactiveTextField.

I solved that problem with a simple **FormGroup** validator.

**Comment by kuhnroyal** [comment]
@joanpablo I would also love to see this feature. I am generating mixins on `FormGroup`s and companion data models based on some form definition. The validators may be different depending on values in the model but the rest of the form stays the same, so I don't want to create 2 different FormGroups.
Since all the form controls are automatically generated in a type-safe manner, I need a way to add validators afterwards.

I would love to override the `FormGroup.addAll(Map<String, AbstractControl<dynamic>> controls)` function in my mixin and just add validators to controls , like so:

```dart
mixin FooForm on FormGroup {
  FormControl<FooEntity> get foo;

  FormControl<String> get name;

  FormControl<String> get description;

  @override
  void addAll(Map<String, AbstractControl<dynamic>> controls) {
    controls['foo'].addValidators([Validators.required]);
    controls['name'].addValidators([Validators.required, Validators.maxLength(42)]);
    super.addAll(controls);
  }
}
```
The build_runner generator actually generates the getters for the form like this:
```dart
  @override
  FormControl<String> get name => control('name) as FormControl<String>;
```

**Comment by joanpablo** [comment]
Hi,

Ok, You have convinced me :smile: I will implement the add validator. @fchiumeo I know that you have to make a PR but I need to do this by myself, and later I will need all of you to help me to test and see if all the needs are covered.

Thanks to all of you.
Best regards

**Comment by joanpablo** [comment]
Hi @kuhnroyal @fchiumeo @piuskamil,

What methods do you think should be implemented:

1-) addValidators: add several validators to the existing ones
2-) setValidators: add validators to the control and override all validators with the new ones.
3-) clearValidators: remove all validators

please let me know what methods to implement

Thanks, best regards

**Comment by kuhnroyal** [comment]
I think `set` and `clear` is enough. I played around with the existing PR, one problem I noticed is, that ` updateValueAndValidity` can not update children (like in Angular?). That currently makes it hard to update several controls in on FormGroup.

**Comment by fchiumeo** [comment]
I agree with @kuhnroyal 
If not, the addValidators method would have to overwrite the validators, because it could have the case: max (10), max (5), max (25) for the same control

**Comment by kuhnroyal** [comment]
> I played around with the existing PR, one problem I noticed is, that ` updateValueAndValidity` can not update children (like in Angular?).

I had a look at the angular sources regarding my earlier comment.
We could use something like this together with the new validator functions:

```ts
 _updateTreeValidity(opts: {emitEvent?: boolean} = {emitEvent: true}) {
    this._forEachChild((ctrl: AbstractControl) => ctrl._updateTreeValidity(opts));
    this.updateValueAndValidity({onlySelf: true, emitEvent: opts.emitEvent});
  }
```

I am adding this as an extension on FormGroup for now to get things working but would be great to have this and call it after adding/setting multiple validators.

**Comment by kuhnroyal** [comment]
This seems to do the job for now:

```dart

extension FormGroupUpdates on FormGroup {
  void updateTreeValueAndValidity({
    bool emitEvent = true,
  }) {
    controls.forEach((name, ctrl) {
      if (ctrl is FormGroup) {
        ctrl.updateTreeValueAndValidity(emitEvent: emitEvent);
      } else {
        ctrl.updateValueAndValidity(updateParent: false, emitEvent: emitEvent);
      }
    });
    updateValueAndValidity(emitEvent: emitEvent);
  }
}
```

**Comment by joanpablo** [comment]
Hi @kuhnroyal,

I'm not sure why you want to trigger validation on children, since you are just adding/removing *validators* to actual control (formGroup, formArray) not to all of its children. ????

**Comment by joanpablo** [comment]
Validators at FormGroup level will maybe take into account value from the FormGroup (implicitly value from children) so you don't need to trigger again all the validators of the children

**Comment by joanpablo** [comment]
The validity of the parent depends on the children, but not the opposite way. That means that when you update the validity status of control you need to update the parent, but no need to update children.

**Comment by kuhnroyal** [comment]
I am adding validators mostly to children of a `FormGroup`, sometimes on the `FormGroup` itself. And sometimes on a `FormGroup` inside a `FormGroup`. Unless I update each of the controls that have had validators changed, the overall validity of the parent is not correct.

See the example above: https://github.com/joanpablo/reactive_forms/issues/129#issuecomment-815636633
I need to call `updateValueAndValidity` for `foo` and for `name`. Or I can just call `this.updateTreeValueAndValidity()`.

I am now doing something like this:
```dart
mixin FooForm on FormGroup {
  FormControl<FooEntity> get foo;

  FormControl<String> get name;

  FormControl<String> get description;

  @override
  void addAll(Map<String, AbstractControl<dynamic>> controls) {
    super.addAll(controls);
    foo.setValidators([Validators.required]);
    name.setValidators([Validators.required, Validators.maxLength(42)]);
    // ... lots more validators
    updateTreeValueAndValidity();
  }
}
```

**Comment by joanpablo** [comment]
Hi, @kuhnroyal @fchiumeo @piuskamil 

The new version released 10.1.0 with set and clear validators dynamically.

Thanks for all your help. Please test it and continue supporting Reactive Forms, all of you are great developers.

Best Regards

**Comment by joanpablo** [comment]
Hi @kuhnroyal,

I have released v10.2.0. In this release you can **setValidators** and also specify an extra argument **autoValidate** for recalculating the new validity status of the control without explicitly call *updateValueAndValidity*.

Please take a look
